openmmtools/integrators.py

```diff
@@ -748,3 +748,144 @@ def __init__(self, temperature=298.0 * simtk.unit.kelvin, collision_rate=91.0 /
+
+    We can then construct integrators by solving each part for a certain timestep in sequence.
+
+    Examples
+    --------
+    >>> ABO_integrator = LangevinSplittingIntegrator("ABO")
+    >>> BAOAB_integrator = LangevinSplittingIntegrator("BAOAB")
+    >>> ABOBA_integrator = LangevinSplittingIntegrator("ABOBA")
+
+    TODO
+    -----
+    - Expressiveness: Currently doesn't allow the user to split up the 'O' step
+
+    References
+    ----------
+    Benedict Leimkuhler, Charles Matthews, 2012
+    Rational Construction of Stochastic Numerical Methods for Molecular Sampling
+    https://arxiv.org/abs/1203.5428
+    '''
+
+    def __init__(self, splitting='BAOAB', temperature=298.0 * simtk.unit.kelvin,
+                 collision_rate=91.0 / simtk.unit.picoseconds, timestep=1.0 * simtk.unit.femtoseconds,
+                 monitor_work=False, monitor_heat=False, position_constraints=True, velocity_constraints=True):
+        '''
+        Parameters
+        ----------
+        splitting : string
+            Sequence of A, B, O steps executed each timestep
+        temperature : numpy.unit.Quantity compatible with kelvin, default: 298.0*simtk.unit.kelvin
+           Fictitious "bath" temperature
+        collision_rate : numpy.unit.Quantity compatible with 1/picoseconds, default: 91.0/simtk.unit.picoseconds
+           Collision rate
+        timestep : numpy.unit.Quantity compatible with femtoseconds, default: 1.0*simtk.unit.femtoseconds
+           Integration timestep
+        monitor_heat : boolean, default: False
+           Accumulate the heat exchanged with the bath in each step, in the global `heat`
+        monitor_work : boolean, default: False
+           Accumulate the shadow work of each step, in the global `shadow_work`
+        position_constraints : boolean, default: True
+           Apply position constraints after every 'A' step
+        velocity_constraints : boolean, default: True
+           Apply velocity constraints after every 'O' or 'B' step
+        '''
+        # Compute constants
+        kT = kB * temperature
+        gamma = collision_rate
+
+        # Make sure `splitting` string is uppercase, for convenience
+        splitting = splitting.upper()
+
+        # Assert it contains at least one each of A,B,O, and no other characters
+        assert (set(splitting) == set('ABO'))
+
+        # Count how many times each step appears, so we know how big each A/B/O substep will be
+        n_A = sum([letter == 'A' for letter in splitting])
+        n_B = sum([letter == 'B' for letter in splitting])
+        n_O = sum([letter == 'O' for letter in splitting])
+
+        # For now, don't split up the O step -- will relax this later!
+        assert (n_O == 1)
+
+        # get strings to pass to self.addComputePerDof(variable, expression) during each A/B/O step
+        update_equations = dict()
+        update_equations['A'] = ("x", "x + (dt / {n_A}) * v".format(n_A=n_A))
```

Comment:
We can probably generalize these update equations to include the potential for multiple lines so we can easily add constraints.